Update child modules docs (#11198)
PyTorchLightning/pytorch-lightning: docs/source/common/lightning_module.rst

        optimizer.step()

> Train Epoch-Level Metrics
> =========================

Reviewer: ```suggestion
Epoch-Level Metrics during Training
==========================
```

This title is a bit confusing

Reviewer: Might be better to have Epoch-Level Metrics and sub section for training, validation and test.

Author: Right now ste structure is like:
```
- Training
    - ...
    - epoch level metrics
    - ...
- Validation
    - ...
    - epoch level metrics
    - ...
- Testing
    - ... 
```
which I think is better because we have other subsections as well that belongs to each of the main heading.

Reviewer 2: @rohitgr7 Could you update the section titles?

.. code-block:: python

    outs = []
    for batch_idx, batch in enumerate(train_dataloader):
        # forward
        out = training_step(batch, batch_idx)

>         outs.append(out)

Author: I swear, I read this example twice 😄

.. code-block:: python

    outs = []
    for batch_idx, batch in enumerate(train_dataloader):
        # forward

>         out = training_step(batch, batch_idx)

Reviewer 3: [same as above] `loss.backward()` below needs updating to `out.backward()`.

    trainer = Trainer()
    trainer.validate(model)

.. note::

>     It is recommended to validate on single device since Distributed Training such as DDP internally

Reviewer: Neat ! Let's make this a warning.

----------

*********
Inference
*********

Prediction Loop
===============

> To activate the prediction loop, you may need to override the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method

Reviewer: ```suggestion
By default, the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method runs under the hook the  :meth:`~pytorch_lightning.core.lightning.LightningModule.forward`. In order to customize this behaviour, simply overrides the  :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method.
```

It might be worth to add an example with `MCDropout` where predict_step is overridden for example

since Lightning under the hood do a forward pass through your LightningModule by default.

.. code-block:: python

    class LitModel(pl.LightningModule):
        def predict_step(self, batch, batch_idx):

>             x, y = batch

Reviewer: Let's show a MCDropout example here and a link to the method usage.

Author: done!

    trainer = Trainer()
    trainer.test(model, dataloaders=test_dataloader)

.. note::

    It is recommended to test on single device since Distributed Training such as DDP internally

>     uses :class:`~torch.utils.data.distributed.DistributedSampler` which replicates some samples to

Reviewer 2: > replicated some samples?

It could worded better.

    make sure all devices have same batch size in case of uneven inputs. This is helpful to make sure
    benchmarking for research papers is done the right way.

----------

*********
Inference
*********

Prediction Loop
===============

By default, the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method runs the
:meth:`~pytorch_lightning.core.lightning.LightningModule.forward` method. In order to customize this behaviour,
simply override the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method.

> For the example let's override ``predict_step`` and try out `Monte Carlo Dropout <https://arxiv.org/pdf/1506.02142.pdf>`_:

Reviewer: Neat !

.. code-block:: python

>     class LitModel(pl.LightningModule):

Reviewer: ```suggestion
    class LitMCdropoutModel(pl.LightningModule):
```

>         def __init__(self):
>             super().__init__()
>             self.linear = nn.Linear(4, 5)

Reviewer: ```suggestion
        def __init__(self, model, mc_iteration: int):
            super().__init__()
            self.linear = model
            self.mc_iteration = mc_iteration
```

            self.dropout = nn.Dropout()

        def predict_step(self, batch, batch_idx):
            # enable Monte Carlo Dropout

>             self.dropout.train()

Reviewer: ```suggestion
            self.train()
```

Author: dropout is the only one that needs to be kept in training mode.

            # take average of 7 iterations

>             pred = torch.vstack([self.dropout(self.linear(x)).unsqueeze(0) for _ in range(7)]).mean(dim=0)

Reviewer: ```suggestion
            pred = torch.vstack([self.dropout(self.linear(x)).unsqueeze(0) for _ in range(self.mc_iteration)]).mean(dim=0)
```

Reviewer: Usually, it is more some layers, dropout, layer, ...

            return pred

Under the hood, Lightning does the following (pseudocode):

.. code-block:: python

    # disable grads + batchnorm + dropout
    torch.set_grad_enabled(False)
    model.eval()
    all_preds = []

>     for batch in predict_dataloader:
>         pred = model.predict_step()

Reviewer: ```suggestion
    for batch_idx, batch in enumerate(predict_dataloader):
        pred = model.predict_step(batch, batch_idx)
```

    multi-device setting, samples could occur duplicated when :class:`~torch.utils.data.distributed.DistributedSampler`
    is used, for eg. with ``strategy="ddp"``. It replicates some samples on some devices to make sure all devices have
    same batch size in case of uneven inputs.

----------

*********
Inference
*********

Prediction Loop
===============

By default, the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method runs the
:meth:`~pytorch_lightning.core.lightning.LightningModule.forward` method. In order to customize this behaviour,
simply override the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method.

For the example let's override ``predict_step`` and try out `Monte Carlo Dropout <https://arxiv.org/pdf/1506.02142.pdf>`_:

.. code-block:: python

    class LitMCdropoutModel(pl.LightningModule):
        def __init__(self, model, mc_iteration):
            super().__init__()
            self.model = model
            self.dropout = nn.Dropout()
            self.mc_iteration = mc_iteration

        def predict_step(self, batch, batch_idx):
            # enable Monte Carlo Dropout
            self.dropout.train()

>             # take average of 7 iterations

Reviewer 3: ```suggestion
            # take average of `self.mc_iteration` iterations
```

    with torch.no_grad():
        y_hat = model(x)

-----------

*************

> Child Modules

Reviewer: I am not sure we want to advise this anymore as it is better to follow the design pattern to simplify using nn.Module without Lightning.

Author: I think it's okay to have as long as all are under `nn.Module`. It represents that LightningModule is just ``nn.Module`` and one can bring modules from outside.

.. code-block:: python

    class Encoder(torch.nn.Module):
        pass

    class Decoder(torch.nn.Module):
        pass

    class AutoEncoder(LightningModule):
        def __init__(self):
            super().__init__()
            self.encoder = Encoder()
            self.decoder = Decoder()
            self.metric = MSE()

        def forward(self, x):
            return self.encoder(x)

        def training_step(self, batch, batch_idx):
            x, _ = batch

            representation = self.encoder(x)
            x_hat = self.decoder(representation)

            loss = self.metric(x, x_hat)
            return loss

        def validation_step(self, batch, batch_idx):
            self._shared_eval(batch, batch_idx, "val")

        def test_step(self, batch, batch_idx):
            self._shared_eval(batch, batch_idx, "test")

        def _shared_eval(self, batch, batch_idx, prefix):
            x, _ = batch
            representation = self.encoder(x)
            x_hat = self.decoder(representation)

            loss = self.metric(x, x_hat)
            self.log(f"{prefix}_loss", loss)

> and we can train this using the same Trainer instance:

Reviewer 4: ```suggestion
and we can train this using the same ``Trainer`` instance:
```

.. code-block:: python

    autoencoder = AutoEncoder()
    trainer = Trainer()
    trainer.fit(autoencoder, train_dataloader, val_dataloader)

> And remember that the forward method should define the practical use of a LightningModule.

Reviewer 4: ```suggestion
And remember that the forward method should define the practical use of a ``LightningModule``.
```

> In this case, we want to use the `AutoEncoder` to extract image representations

Reviewer 4: ```suggestion
In this case, we want to use the ``AutoEncoder`` to extract image representations
```

        if self.current_epoch == 0:
            ...

> -------------

Reviewer 5: Are we removing these because we are  updating the styling and spacing globally in the docs?

So far @williamFalcon  strongly insisted in this particular formatting style, so we should keep it unless clearly agreed on.

Author: yes.. mostly the top-level headings are separated by these breaks. Adding it to all of these individual properties adds up a lot of space/padding in between. Without them, it looks good to me.
<img width="875" alt="Screenshot 2022-01-10 at 1 46 23 AM" src="https://user-images.githubusercontent.com/30778939/148699206-b4ad5137-e955-4291-bc70-55c3a6115f19.png">

        # generate some images using the example_input_array
        gen_images = self.generator(self.example_input_array)

> datamodule

Reviewer: This is deprecated right ?

Author: yes..thanks :)
